onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -49,13 +51,30 @@ std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_DML(ID
   return std::make_shared<onnxruntime::DMLProviderFactory>(dml_device, cmd_queue);
 }
 
+bool IsSoftwareAdapter(IDXGIAdapter1* adapter)
+{
+    DXGI_ADAPTER_DESC1 pDesc;
+    adapter->GetDesc1(&pDesc);
+
+    // see here for documentation on filtering WARP adapter:
+    // https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/d3d10-graphics-programming-guide-dxgi#new-info-about-enumerating-adapters-for-windows-8
+    auto isBasicRenderDriverVendorId = pDesc.VendorId == 0x1414;
+    auto isBasicRenderDriverDeviceId = pDesc.DeviceId == 0x8c;
+    auto isSoftwareAdapter = pDesc.Flags == DXGI_ADAPTER_FLAG_SOFTWARE;
+    
+    return isSoftwareAdapter || (isBasicRenderDriverVendorId && isBasicRenderDriverDeviceId);
+}
+
 std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_DML(int device_id) {
   ComPtr<IDXGIFactory4> dxgi_factory;
   THROW_IF_FAILED(CreateDXGIFactory2(0, IID_PPV_ARGS(&dxgi_factory)));
 
   ComPtr<IDXGIAdapter1> adapter;
   THROW_IF_FAILED(dxgi_factory->EnumAdapters1(device_id, &adapter));
 
+  // Disallow using DML with the software adapter (Microsoft Basic Display Adapter) because CPU evaluations are much faster
+  THROW_HR_IF(E_INVALIDARG, IsSoftwareAdapter(adapter.Get()));
```

Comment:
In Windows "Device Manager", I saw each our build machine has two display adapters:
1. The tesla
2. A basic display adapter

I guess both of them will be returned from the "EnumAdapters1" function?
Is it possible that sometimes the device of id 0 is the tesla but sometimes it is the basic one?

Reply:
The hardware device will be enumerated first